### PR TITLE
Support deploying with docker compose without local docker-compose

### PR DESCRIPTION
**Pull request by fiftin**
Related to this issue: https://github.com/testcontainers/testcontainers-go/issues/425.

I added support conteinerised docker-compose. To do this I refectored `compose.go` and extracted `ComposeExecutor` which can be:
* `ContainerizedDockerComposeExecutor` - uses docker-compose from docker container.
* `LocalDockerComposeExecutor` - uses local docker-compose.

### CODE REVIEW

File: compose.go [hunk 1]
```diff
@@ -76,29 +75,25 @@ func (f LocalDockerComposeOptionsFunc) ApplyToLocalCompose(opts *LocalDockerComp
...
 
-// NewLocalDockerCompose returns an instance of the local Docker Compose, using an
-// array of Docker Compose file paths and an identifier for the Compose execution.
-//
-// It will iterate through the array adding '-f compose-file-path' flags to the local
-// Docker Compose execution. The identifier represents the name of the execution,
-// which will define the name of the underlying Docker network and the name of the
-// running Compose services.
-func NewLocalDockerCompose(filePaths []string, identifier string, opts ...LocalDockerComposeOption) *LocalDockerCompose {
+func NewDockerCompose(
```
Comment (mdelapenya) [inline compose.go:78]: This signature change comes with a breaking change :) Thankfully we are not in 1.x version, so we can change contracts

Comment (mdelapenya) [reply compose.go:78]: Oh oh wait. Sorry, but I thought this line was changing the `NewLocalDockerCompose` method, but it's not: it's adding a new generic method.

File: compose_executors.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	if len(cmds) == 0 {
+		cmds = append(cmds, "-f", "/app/docker-compose.yml")
+	}
+
+	args := append(cmds, options.Args...)
+
+	req := ContainerRequest{
+		Image: "docker/compose:1.29.2",
```
[the remark this answers is not in the document]
Comment (fiftin) [reply compose_executors.go:64]: Agree, will do

File: compose.go [hunk 3]
```diff
@@ -116,6 +111,21 @@ func NewLocalDockerCompose(filePaths []string, identifier string, opts ...LocalD
...
+//
+// It will iterate through the array adding '-f compose-file-path' flags to the local
+// Docker Compose execution. The identifier represents the name of the execution,
+// which will define the name of the underlying Docker network and the name of the
+// running Compose services.
+func NewLocalDockerCompose(filePaths []string, identifier string, opts ...LocalDockerComposeOption) *LocalDockerCompose {
+	return NewDockerCompose(filePaths, identifier, LocalDockerComposeExecutor, nil, opts...)
+}
+
+func NewContainerizedDockerCompose(filePaths []string, identifier string, opts ...LocalDockerComposeOption) *LocalDockerCompose {
```
Comment (mdelapenya) [inline compose.go:125]: Do we still to return a `Local`DockerCompose? It seems weird to me to use a containerised compose method that returns a LocalDC.

I'd be open to using a dedicated DockerisedDockerCompose type, although I think could come with changes in return types and generalisation...

Of course, not a blocker for this PR, but if you are interested in following-up with more refactors to make a stable API, I'm too

File: compose_test.go [hunk 4]
```diff
@@ -90,68 +89,91 @@ func ExampleLocalDockerCompose_WithEnv() {
 	})
 }
 
+func enumDockerComposes(filePaths []string, identifier string, executor func(compose *LocalDockerCompose), t *testing.T) {
+	composes := []*LocalDockerCompose{
+		NewLocalDockerCompose(filePaths, identifier, WithLogger(TestLogger(t))),
+		NewContainerizedDockerCompose(filePaths, identifier, WithLogger(TestLogger(t))),
```
Comment (mdelapenya) [inline compose_test.go:95]: I think we are using the new Dockerised compose here. Could you update the docs with examples on how to use it? There is a dedicated section for compose in the docs/features dir.

File: compose.go (hunk 3 above)
Comment (fiftin) [reply compose.go:125]: I started with creating dedicated `DockerisedDockerCompose` type, but got too many duplications. Using current `LocalDockerCompose` with different executors is a best solution for my opinion.

IMHO: interface `DockerCompose` should be removed and type `LocalDockerCompose` should renamed to `DockerCompose`.

P.S.:

I replaced ref to function to interface `DockerComposeExecutor` with 2 implementations.

File: compose_executors.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fiftin) [reply compose_executors.go]: I replaced ref to function to interface `DockerComposeExecutor` with 2 implementations.

**Comment by fiftin** [comment]
I replaced ref to function to interface `DockerComposeExecutor` with 2 implementations.

File: compose_test.go (hunk 4 above)
Comment (fiftin) [reply compose_test.go:95]: Done

**Pull request closed without merging**